src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs

```diff
@@ -0,0 +1,979 @@
+            Interop.NtDll.CreateDisposition dwCreateDisposition = 0;
+            System.IO.FileShare nShareAccess = System.IO.FileShare.None;
+
+            switch (access)
+            {
+                case System.IO.FileAccess.Read:
+
+                    nDesiredAccess |= Interop.NtDll.DesiredAccess.FILE_READ_DATA;
+                    nShareAccess = System.IO.FileShare.Delete | System.IO.FileShare.ReadWrite;
+                    dwCreateDisposition = Interop.NtDll.CreateDisposition.FILE_OPEN;
+                    break;
+
+                case System.IO.FileAccess.Write:
+                    nDesiredAccess |= Interop.NtDll.DesiredAccess.FILE_WRITE_DATA;
+                    nShareAccess = System.IO.FileShare.Delete | System.IO.FileShare.Read;
+                    dwCreateDisposition = Interop.NtDll.CreateDisposition.FILE_OVERWRITE;
+                    break;
+
+                case System.IO.FileAccess.ReadWrite:
+                default:
+                    // we validate the value of 'access' parameter in the beginning of this method
+                    Debug.Assert(access == System.IO.FileAccess.ReadWrite);
+
+                    nDesiredAccess |= Interop.NtDll.DesiredAccess.FILE_READ_DATA | Interop.NtDll.DesiredAccess.FILE_WRITE_DATA;
+                    nShareAccess = System.IO.FileShare.Delete | System.IO.FileShare.Read;
+                    dwCreateDisposition = Interop.NtDll.CreateDisposition.FILE_OVERWRITE;
+                    break;
+            }
+
+            if ((options & System.IO.FileOptions.WriteThrough) != 0)
+            {
+                dwCreateOptions |= Interop.NtDll.CreateOptions.FILE_WRITE_THROUGH;
+            }
+
+            if ((options & System.IO.FileOptions.Asynchronous) == 0)
+            {
+                dwCreateOptions |= Interop.NtDll.CreateOptions.FILE_SYNCHRONOUS_IO_NONALERT;
+            }
+
+            if ((options & System.IO.FileOptions.SequentialScan) != 0)
+            {
+                dwCreateOptions |= Interop.NtDll.CreateOptions.FILE_SEQUENTIAL_ONLY;
+            }
+
+            if ((options & System.IO.FileOptions.RandomAccess) != 0)
+            {
+                dwCreateOptions |= Interop.NtDll.CreateOptions.FILE_RANDOM_ACCESS;
+            }
+
+            try
+            {
+                eaBuffer = new FileFullEaInformation(transactionContext);
+
+                // NOTE: the Name property is intended to reveal the publicly available moniker for the
+                // FILESTREAM attributed column data. We will not surface the internal processing that
+                // takes place to create the mappedPath.
+                string mappedPath = InitializeNtPath(sPath);
+                int retval = 0;
+                Interop.Kernel32.SetThreadErrorMode(Interop.Kernel32.SEM_FAILCRITICALERRORS, out uint oldMode);
+
+                try
+                {
+
+                    (int status, IntPtr handle) = Interop.NtDll.CreateFile( 
```

Comment:
Why do we need to use Win32 interop for this?  i.e. why can't this just use FileStream?

Reply:
As per reasons mentioned in #29379, since it is not a real UNC path, the bogus path only has meaning in the context of this transaction, hence we cannot use FileStream directly.